crates/sui-core/src/execution_cache/writeback_cache.rs

```diff
@@ -913,6 +950,44 @@ impl WritebackCache {
                 .is_some());
             self.flush_transactions_from_dirty_to_cached(epoch, *tx_digest, outputs);
         }
+
+        let num_outputs = all_outputs.len() as u64;
+        let num_commits = self
+            .dirty
+            .total_transaction_commits
+            .fetch_add(num_outputs, std::sync::atomic::Ordering::Relaxed)
+            + num_outputs;
+
+        let pending_count = self
+            .dirty
+            .total_transaction_inserts
+            .load(std::sync::atomic::Ordering::Relaxed)
+            .saturating_sub(num_commits);
+
+        self.set_backpressure(pending_count);
+    }
+
+    fn approximate_pending_transaction_count(&self) -> u64 {
+        let num_commits = self
+            .dirty
+            .total_transaction_commits
+            .load(std::sync::atomic::Ordering::Relaxed);
+
+        self.dirty
+            .total_transaction_inserts
+            .load(std::sync::atomic::Ordering::Relaxed)
+            .saturating_sub(num_commits)
+    }
+
+    fn set_backpressure(&self, pending_count: u64) {
```

Comment:
My understanding of atomic ordering primitives is rusty to say the least

Given the use of `Relaxed` everywhere, what prevents us from getting into a deadlock where backpressure is stuck on, because commits are not observed and new tx are not coming in? (sorry if this question is invalid. this is approximately the level of specificity I am able to achieve in describing my concern, after reading the c++ memory ordering reference for 5 minutes)

is this where the "suppression" thing comes into play?

Reply:
I'm not sure I understand the question - if there are no commits arriving then it doesn't matter if there is backpressure or not, since the main thing backpressure would do is pause commit handling.

But under normal circumstances backpressure would be lifted because checkpoint execution will flush transactions to disk and increment the num_commits counter.

The use of relaxed shouldn't matter here because there is only a single memory location that is observed by both threads. Stricter orderings are necessary when one thread cares about the ordering of more than one event that occurred on a different thread, i.e. a memory barrier / fence is required